File: mini-claw-code-py/src/mini_claw_code_py/tools/write.py
from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any

from ..types import ToolDefinition


class WriteTool:
    def __init__(self) -> None:
        self._definition = (
            ToolDefinition.new(
                "write",
                "Write content to a file, creating directories as needed.",
            )
            .param("path", "string", "The file path to write to", True)
            .param("content", "string", "The content to write to the file", True)
        )

    @property
    def definition(self) -> ToolDefinition:
        return self._definition

    async def call(self, args: Any) -> str:
        path = args.get("path") if isinstance(args, dict) else None
        content = args.get("content") if isinstance(args, dict) else None
        if not isinstance(path, str):
            raise ValueError("missing 'path' argument")
        if not isinstance(content, str):
            raise ValueError("missing 'content' argument")

        path_obj = Path(path)
        if path_obj.parent:
            await asyncio.to_thread(path_obj.parent.mkdir, parents=True, exist_ok=True)
        try:
            await asyncio.to_thread(path_obj.write_text, content)
        except Exception as exc:
            raise RuntimeError(f"failed to write '{path}'") from exc
        return f"wrote {path}"

File: mini-claw-code-ts/src/tools/write.ts
import { mkdir, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import { ToolDefinition, type JsonObject, type Tool } from "../types";

export class WriteTool implements Tool {
  readonly #definition = new ToolDefinition(
    "write",
    "Write content to a file, creating directories as needed.",
  )
    .param("path", "string", "The file path to write to", true)
    .param("content", "string", "The content to write to the file", true);

  definition(): ToolDefinition {
    return this.#definition;
  }

  async call(args: JsonObject | null): Promise<string> {
    const path = args?.path;
    const content = args?.content;

    if (typeof path !== "string") {
      throw new Error("missing 'path' argument");
    }
    if (typeof content !== "string") {
      throw new Error("missing 'content' argument");
    }

    await mkdir(dirname(path), { recursive: true });
    await writeFile(path, content, "utf8");
    return `wrote ${path}`;
  }
}
